pyomo/contrib/community_detection/community_graph.py

```diff
@@ -0,0 +1,343 @@
+            if node not in constraint_variable_map:
+                connected_constraints = []  # Initialize list of constraints that share this variable
+
+                # Loop through the variable node's edges to find constraints that contain the variable
+                for edge in bipartite_model_graph.edges(node):
+                    # The first node in the edge tuple will always be the node that is used as the
+                    # argument in 'bipartite_model_graph.edges(node)'
+                    # Thus, the relevant node is the second one in the tuple 'edge[1]'
+                    connected_constraints.append(edge[1])
+
+                # Create all possible two-node combinations from connected_constraints; in other words, create a list
+                # of all the edges that need to be created between constraints based on connected_constraints
+                edges_between_constraints = list(combinations(sorted(connected_constraints), 2))
+
+                # Update edge_weight_dict or edge_set based on the determined edges between nodes
+                if weighted_graph:
+                    new_edge_weights = {edge: edge_weight_dict.get(edge, 0) + 1 for edge in edges_between_constraints}
+                    edge_weight_dict.update(new_edge_weights)
+                else:
+                    edge_set.update(edges_between_constraints)
+
+            # If the node is not a variable, then it is a constraint and we want it as
+            # a node in our constraint node graph
+            else:
+                collapsed_model_graph.add_node(node)
+
+    elif type_of_graph == 'v':  # Variable node graph - collapse the bipartite graph into a variable node graph
+
+        for node in bipartite_model_graph.nodes():
+
+            # If the node represents a constraint (or objective)
+            if node in constraint_variable_map:
+
+                # Create list of variables in the constraint equation
+                connected_variables = constraint_variable_map[node]
+
+                # Create all possible two-node combinations from connected_variables; in other words, create a list
+                # of all the edges that need to be created between variables based on connected_variables
+                edges_between_variables = list(combinations(sorted(connected_variables), 2))
+
+                # Update edge_weight_dict or edge_set based on the determined edges_between_nodes
+                if weighted_graph:
+                    new_edge_weights = {edge: edge_weight_dict.get(edge, 0) + 1 for edge in edges_between_variables}
+                    edge_weight_dict.update(new_edge_weights)
+                else:
+                    edge_set.update(edges_between_variables)
+
+            # If the node is not a constraint/objective, then it is a variable and we want it as
+            # a node in our variable node graph
+            else:
+                collapsed_model_graph.add_node(node)
```

Comment:
First thought: these should be a separate private functions.

Second thought: does networkx have a functionality for this type of graph operation? I would presume so. If so, I would use that.

Reply:
You are absolutely right. This was an oversight from me. There is a function `weighted_projected_graph` in NetworkX that does this exactly (see https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.bipartite.projection.weighted_projected_graph.html#networkx.algorithms.bipartite.projection.weighted_projected_graph). I will discuss this with Rahul and try to change this implementation